doc: custom board programming guide (#19750)
nrfconnect/sdk-nrf: doc/nrf/app_dev/device_guides/custom/programming_custom_board.rst

.. _programming_custom_board:

Connecting custom boards for programming
########################################

.. contents::
   :local:
   :depth: 2

After you :ref:`created custom board files <defining_custom_board>`, you need to connect your custom board to a debug probe for programming.

Hardware requirements
*********************

To connect a custom board for programming, you need the following hardware:

* A debug probe supporting Serial Wire Debug (SWD), such as:

  * A Development Kit (DK) from Nordic Semiconductor (all include an onboard debug probe)
  * Dedicated debug adapter (like SEGGER J-Link)

* Connecting wires

> * USB cable

Reviewer: ```suggestion
* USB cable for DK or debug adapter
```

Required connections
********************

> You will have to connect the following pins between your custom board and the debug probe:

Reviewer 2: ```suggestion
You have to connect the following pins between your custom board and the debug probe:
```

* Essential connections:

  * SWDCLK (Clock)
  * SWDIO (Data)

>   * GND (Ground)

Reviewer: I think we also need Voltage Detect, or whatever it is named, so the debugger knows which voltage to use for programming

* Optional connections:

  * VDD (for powering the board from debug probe)
  * RESET

> 

Reviewer: We should also say that if the custom board is not powered from VDD, it must be powered externally. This is a common thing for users to miss.

Connecting the debug probe
**************************

The following steps describe how to connect your custom board to a debug probe.

.. tabs::

   .. group-tab:: Using a DK as debug probe

      Development Kits from Nordic Semiconductor include an onboard debug probe that supports the J-Link interface.

      To connect your custom board to this onboard debug probe, complete the following steps:

      1. Locate the SWD debug output header pins on your DK.
         Check the DK user guide on `Nordic Semiconductor TechDocs`_ for the exact location of these pins.
         For example, for the nRF52840 DK, read the `Debug output <nRF52840 DK Debug output_>`_ page.
      #. Connect the required pins to your custom board.
      #. Connect the DK to your PC using the USB cable.
      #. Install required J-Link drivers if not already present.

   .. group-tab:: Using a dedicated debug adapter

>       If you don't have a Development Kit, use a dedicated debug adapter:

Reviewer 2: ```suggestion
      If you do not have a Development Kit, use a dedicated debug adapter:
```

>       1. Connect a compatible debug probe (like SEGGER J-Link) to your custom board.

Reviewer: Maybe we should say that they should follow debugger docs on how/what to connect?

Connecting the debug probe
**************************

The following steps describe how to connect your custom board to a debug probe.

.. tabs::

   .. group-tab:: Using a DK as debug probe

      Development Kits from Nordic Semiconductor include an onboard debug probe that supports the J-Link interface.

      To connect your custom board to this onboard debug probe, complete the following steps:

      1. Locate the SWD debug output header pins on your DK.
         Check the DK user guide on `Nordic Semiconductor TechDocs`_ for the exact location of these pins.
         For example, for the nRF52840 DK, read the `Debug output <nRF52840 DK Debug output_>`_ page.
      #. Connect the required pins to your custom board.
      #. Connect the DK to your PC using the USB cable.
      #. Install required J-Link drivers if not already present.

   .. group-tab:: Using a dedicated debug adapter

      If you do not have a Development Kit, use a dedicated debug adapter:

      1. Connect a compatible debug probe (like SEGGER J-Link) to your custom board.
         See the debug probe documentation for the exact pinout.

>       2. Install the appropriate debug interface drivers.
>       3. Connect the debug probe to your PC using the USB cable.

Reviewer 3: ```suggestion
      #. Install the appropriate debug interface drivers.
      #. Connect the debug probe to your PC using the USB cable.
```
